File: ProjectOne/src/testing/ContactServiceTest.java
/*
 * Jeremy Cameron
 * CS 320
 * ContactServiceTest class which is used for conducting unit tests for the ContactService class.
 */

package testing;

import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import contacts.Contact;
import contacts.ContactService;

public class ContactServiceTest {
	
	// Instance of ContactService used in the unit tests
	private ContactService contactService;
	
	/*
	 * Initializes a new instance of ContactService before each test.
	 * This makes sure that each test runs with a fresh instance of ContactService.
	 */
	@BeforeEach
	public void setUp() {
		contactService = new ContactService();
	}
	
	/*
	 * Tests adding a new contact to the service.
	 * Verifies that a contact can be retrieved after being added.
	 */
	@Test
	public void testAddContact() {
		Contact contact = new Contact("Jeremy", "Cameron", "1235157298", "1111 Your St");
		contactService.addContact(contact);
		assertNotNull(contactService.getContact(contact.getId()));
	}
	
	/*
	 * Tests adding a duplicate contact to the service.
	 * Verifies that an IllegalArgumentException is thrown when trying to add a duplicate contact ID.
	 */
	@Test
	public void testAddDuplicateContact() {
		Contact contact1 = new Contact("Jeremy", "Cameron", "1235157298", "1111 Your St");
		Contact contact2 = new Contact("Jamie", "Cameron", "2226117828", "2222 My Ave");
		contactService.addContact(contact1);
		contactService.addContact(contact2);
		assertThrows(IllegalArgumentException.class, () -> {
			contactService.addContact(contact1);
		});
	}
	
	/*
	 * Tests deleting a contact from the service.
	 * Verifies that a contact can be deleted and that an IllegalArgumentException is thrown when 
	 * trying to retrieve the contact ID afterwards.
	 */
	@Test
	public void testDeleteContact() {
		Contact contact = new Contact("Jeremy", "Cameron", "1235157298", "1111 Your St");
		contactService.addContact(contact);
		String contactId = contact.getId();
		//System.out.println("Attempting to delete contact with ID: " + contactId);
		contactService.deleteContact(contactId);
		assertThrows(IllegalArgumentException.class, () -> {
			contactService.getContact(contactId);
		});
	}
	
	/*
	 * Tests trying to delete a contact ID that does not exist.
	 * Verifies that an IllegalArgumentException is thrown when trying to delete a contact 
	 * that does not exist.
	 */
	@Test
	public void testDeleteInvalidContact() {
		String invalidContact = "Contact ID does not exist.";
		assertThrows(IllegalArgumentException.class, () -> {
			contactService.deleteContact(invalidContact);
		});
	}
	
	/*
	 * Tests updating the first name of a contact.
	 * Verifies that the first name can be updated successfully.
	 */
	@Test
	public void testUpdateFirstName() {
		Contact contact = new Contact("Jeremy", "Cameron", "1235157298", "1111 Your St");
		contactService.addContact(contact);
		contactService.updateFirstName(contact.getId(), "Jamie");
		Contact updatedContact = contactService.getContact(contact.getId());
		assertEquals("Jamie", updatedContact.getFirstName());
	}
	
	/*
	 * Tests updating the last name of a contact.
	 * Verifies that the last name can be updated successfully.
	 */
	@Test
	public void testUpdateLastName() {
		Contact contact = new Contact("Jeremy", "Cameron", "1235157298", "1111 Your St");
		contactService.addContact(contact);
		contactService.updateLastName(contact.getId(), "Welker");
		Contact updatedContact = contactService.getContact(contact.getId());
		assertEquals("Welker", updatedContact.getLastName());
	}
	
	/*
	 * Tests updating the phone number.
	 * Verifies that the phone number can be updated successfully.
	 */
	@Test
	public void testUpdatePhoneNumber() {
		Contact contact = new Contact("Jeremy", "Cameron", "1235157298", "1111 Your St");
		contactService.addContact(contact);
		contactService.updatePhoneNumber(contact.getId(), "2226117828");
		Contact updatedContact = contactService.getContact(contact.getId());
		assertEquals("2226117828", updatedContact.getPhoneNumber());
	}
	
	/*
	 * Tests updating the address.
	 * Verifies that the address can be updated successfully.
	 */
	@Test
	public void testUpdateAddress() {
		Contact contact = new Contact("Jeremy", "Cameron", "1235157298", "1111 Your St");
		contactService.addContact(contact);
		contactService.updateAddress(contact.getId(), "2222 My Ave");
		Contact updatedContact = contactService.getContact(contact.getId());
		assertEquals("2222 My Ave", updatedContact.getAddress());
	}
	
	/*
	 * Tests trying to update a contact that does not exist.
	 * Verifies that an IllegalArgumentException is thrown when trying to update a contact
	 * that does not exist.
	 */
	@Test
	public void testUpdateInvalidContact() {
		assertThrows(IllegalArgumentException.class, () -> {
			contactService.updateFirstName("Contact ID does not exist.", "Jamie");
		});
		assertThrows(IllegalArgumentException.class, () -> {
			contactService.updateLastName("Contact ID does not exist.", "Welker");
		});
		assertThrows(IllegalArgumentException.class, () -> {
			contactService.updatePhoneNumber("Contact ID does not exist.", "2226117828");
		});
		assertThrows(IllegalArgumentException.class, () -> {
			contactService.updateAddress("Contact ID does not exist.", "2222 My Ave");
		});
	}
	
	/*
	 * Tests updating a contact with an invalid phone number.
	 * Verifies that an IllegalArgumentException is thrown when trying to update the contact
	 * with an invalid phone number. 
	 */
	@Test
	public void testUpdateInvalidPhoneNumber() {
		Contact contact = new Contact("Jeremy", "Cameron", "1235157298", "1111 Your St");
		contactService.addContact(contact);
		assertThrows(IllegalArgumentException.class, () -> {
			contactService.updatePhoneNumber(contact.getId(), "2226");
		});
	}
	
	/*
	 * Tests updating a contact with an invalid address.
	 * Verifies that an IllegalArgumentException is thrown when trying to update the contact
	 * with an invalid address.
	 */
	@Test
	public void testUpdateInvalidAddress() {
		Contact contact = new Contact("Jeremy", "Cameron", "1235157298", "1111 Your St");
		contactService.addContact(contact);
		assertThrows(IllegalArgumentException.class, () -> {
			contactService.updateAddress(contact.getId(), "1111 My Ave, Apt G, city, state, zip code");
		});
	}
}

File: enhancement_1/CS499_Project/src/Tests/contact-service.test.js
// tests/contact-service.test.js
import ContactService from '../services/contact-service.js';

describe('ContactService', () => {
  let contactService;
  let contact;

  // Initializes contactService and creates a test contact before each test
  beforeEach(() => {
    contactService = new ContactService();
    contact = contactService.addContact({
      firstName: 'Jeremy',
      lastName: 'Cameron',
      phoneNumber: '1235157298',
      address: '1111 Your St'
    });
  });

  // Test for successfully adding a new contact
  test('adds a new contact', () => {
    expect(contactService.getContact(contact.id)).toBeDefined();
  });

  // Test for throwing a duplicate ID error when trying to add a new contact with an ID that already exists
  test('throws error when adding duplicate contact', () => {
  const duplicateContact = {
    id: contact.id,
    firstName: 'Jeremy',
    lastName: 'Cameron',
    phoneNumber: '1235157298',
    address: '1111 Your St'
  };

  expect(() => contactService.addContact(duplicateContact)).toThrow();
});

  // Test for successfully deleting a contact
  test('deletes a contact', () => {
    contactService.deleteContact(contact.id);
    expect(() => contactService.getContact(contact.id)).toThrow();
  });

  // Test for throwing an error when trying to delete a contact that is not found
  test('throws error when deleting a contact that is not found', () => {
    expect(() => contactService.deleteContact('Contact not found')).toThrow();
  });

  // Test for sucessfully updating a contact's first name
  test('updates first name', () => {
    contactService.updateFirstName(contact.id, 'Jamie');
    expect(contactService.getContact(contact.id).firstName).toBe('Jamie');
  });

  // Test for successfully updating a contact's last name
  test('updates last name', () => {
    contactService.updateLastName(contact.id, 'Welker');
    expect(contactService.getContact(contact.id).lastName).toBe('Welker');
  });

  // Test for successfully updating a contact's phone number
  test('updates phone number', () => {
    contactService.updatePhoneNumber(contact.id, '2226117828');
    expect(contactService.getContact(contact.id).phoneNumber).toBe('2226117828');
  });

  // Test for successfully updating a contact's address
  test('updates address', () => {
    contactService.updateAddress(contact.id, '2222 My Ave');
    expect(contactService.getContact(contact.id).address).toBe('2222 My Ave');
  });

  // Test for throwing an error when trying to update a contact that is not found
  test('throws error when updating a contact that is not found', () => {
    expect(() => contactService.updateFirstName('Contact not found', 'Jamie')).toThrow();
    expect(() => contactService.updateLastName('Contact not found', 'Welker')).toThrow();
    expect(() => contactService.updatePhoneNumber('Contact not found', '2226117828')).toThrow();
    expect(() => contactService.updateAddress('Contact not found', '2222 My Ave')).toThrow();
  });

  // Test for throwing an error when trying to update a contact with an invalid phone number
  test('throws error when updating with an invalid phone number', () => {
    expect(() => contactService.updatePhoneNumber(contact.id, '2226')).toThrow();
  });

  // Test for throwing an error when trying to update a contact with an invalid address
  test('throws error when updating with an invalid address', () => {
    expect(() => contactService.updateAddress(contact.id, '1111 My Ave, Apt G, city, state, zip code')).toThrow();
  });
});
